### ISSUE TITLE
Incorrect error message in evaluation of wrong perf_pattern

### ISSUE THREAD

**Issue by OliverPerks** [eth-cscs/reframe#2387]
[MWE.zip](https://github.com/eth-cscs/reframe/files/7890023/MWE.zip)

Silly issue with error reporting on wrong code.

```
       pref_regex = r'Hello\s(\S+)'
       times = sn.extractall(pref_regex, self.stdout, 1, int)
       self.perf_patterns = {
               'Total Time': sum(times)/len(times)
       }
```
(I know now that I should be using `sn.avg(times)` but thats not the point here.)

Reframe fails with:

```
$ reframe -r -c hello.py --performance-report
....
[  FAILED  ] Ran 1/1 test case(s) from 1 check(s) (1 failure(s), 0 skipped)
[==========] Finished on Tue Jan 18 08:49:55 2022 

==============================================================================
SUMMARY OF FAILURES
------------------------------------------------------------------------------
FAILURE INFO for HelloTest 
  * Test Description: HelloTest
  * System partition: PACU-login2:default
  * Environment: builtin
  * Stage directory: /home/ollper01/tmp/rframe/stage/PACU-login2/default/builtin/HelloTest
  * Node list: pacu01
  * Job type: local (id=3696722)
  * Dependencies (conceptual): []
  * Dependencies (actual): []
  * Maintainers: []
  * Failing phase: sanity
  * Rerun with '-n HelloTest -p builtin --system PACU-login2:default -r'
  * Reason: sanity error: rfm_HelloTest_job.out: No such file or directory
------------------------------------------------------------------------------

```

So it is reporting that the output file doesn't exist. Which clearly it does, and is not the issue at hand.

Running in a more verbose mode:

```
$ reframe -vvv -r -c hello.py --performance-report
....
==============================================================================
SUMMARY OF FAILURES
------------------------------------------------------------------------------
FAILURE INFO for HelloTest 
  * Test Description: HelloTest
  * System partition: PACU-login2:default
  * Environment: builtin
  * Stage directory: /home/ollper01/tmp/rframe/stage/PACU-login2/default/builtin/HelloTest
  * Node list: pacu01
  * Job type: local (id=3697015)
  * Dependencies (conceptual): []
  * Dependencies (actual): []
  * Maintainers: []
  * Failing phase: sanity
  * Rerun with '-n HelloTest -p builtin --system PACU-login2:default -r'
  * Reason: sanity error: rfm_HelloTest_job.out: No such file or directory
Traceback (most recent call last):
  File "/home/ollper01/Foundry/reframe/reframe/utility/sanity.py", line 37, in _open
    with open(filename, *args, **kwargs) as fp:
FileNotFoundError: [Errno 2] No such file or directory: 'rfm_HelloTest_job.out'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ollper01/Foundry/reframe/reframe/frontend/executors/__init__.py", line 267, in _safe_call
    return fn(*args, **kwargs)
  File "/home/ollper01/Foundry/reframe/reframe/core/hooks.py", line 78, in _fn
    getattr(obj, h.__name__)()
  File "/home/ollper01/Foundry/reframe/reframe/core/hooks.py", line 32, in _fn
    func(*args, **kwargs)
  File "/home/ollper01/tmp/rframe/hello.py", line 31, in set_sanity_patterns
    'Total Time': sum(times)/len(times)
  File "/home/ollper01/Foundry/reframe/reframe/core/deferrable.py", line 94, in __iter__
    return iter(self.evaluate())
  File "/home/ollper01/Foundry/reframe/reframe/core/deferrable.py", line 69, in evaluate
    ret = self._fn(*fn_args, **fn_kwargs)
  File "/home/ollper01/Foundry/reframe/reframe/utility/sanity.py", line 795, in extractall
    for x in extractiter(patt, filename, tag, conv, encoding))
  File "/home/ollper01/Foundry/reframe/reframe/utility/sanity.py", line 794, in <genexpr>
    return list(evaluate(x)
  File "/home/ollper01/Foundry/reframe/reframe/utility/sanity.py", line 735, in extractiter
    with _open(filename, 'rt', encoding=encoding) as fp:
  File "/usr/lib64/python3.6/contextlib.py", line 81, in __enter__
    return next(self.gen)
  File "/home/ollper01/Foundry/reframe/reframe/utility/sanity.py", line 41, in _open
    raise SanityError(f'{filename}: {e.strerror}')
reframe.core.exceptions.SanityError: rfm_HelloTest_job.out: No such file or directory

------------------------------------------------------------------------------
```

Here we can finally spot the actual issue:
```
File "/home/ollper01/tmp/rframe/hello.py", line 31, in set_sanity_patterns
    'Total Time': sum(times)/len(times)
```

However, the `No such file or directory` error is still presented as the issue a number of times, which confused me when debugging this in the real case.

Again, this is low priority, and is a user error - but error messages and debugging are really important when developing test files.
No big deal, but hope this MWE can be useful.

**Comment by mahendrapaipuri** [comment]
Hello @OliverPerks, I think the error message makes sense. The thing is `perf_patterns` must be deferred expressions and they are "evaluated" after job is finished in post run hooks. Your expression `sum(times)/len(times)` will trigger the evaluation of the deferred expression even before the job is submitted to the scheduler. At this point, there is no `stdout` file in the stage directory and that is what ReFrame is complaining about.

**Comment by OliverPerks** [comment]
Thanks for this. Yes that makes some sense.

However, given the uses of deferred expressions then ability debug is even more important.
If this clause had failed before the job has been run (which I can believe) why is the error not reported at that point. In my real case the whole job then gets executed (~15 mins) and then I get the error.

This isn’t a hill I’m willing to die on if you all feel this is an edge case / non serious then that’s fine, but it certainly cost me a few hours, and I wouldn’t like others to fall victim too.

Thanks so much for your help with this.
Olly

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by mahendrapaipuri** [comment]
I was a victim myself once. Yes, I remember that in my case too this error was raised only after finishing the job. I think the ReFrame core development team has to look into this issue. I agree that it would be indeed nicer if ReFrame can detect and raise this error before job submission.

**Comment by vkarak** [comment]
The problem is the following and not related to the expressions being deferred, because at the time this is called, the job has finished and the output *is* produced. You are in a pre-sanity hook, so the job has finished. The problem is that all pipeline hooks execute from the *current working directory*. There is a note in the manual (I'm not sure if is mentioned in the tutorial though) [here](https://reframe-hpc.readthedocs.io/en/stable/regression_test_api.html#pipeline-hooks). I agree that this is a common pitfall, and perhaps we should stress it much more! The reason for not choosing run the hooks from the stage directory is that the stage directory appears after the setup phase, which would mean different types of hooks would execute in different directories and we didn't want that. That's why, specifically for the sanity and performance checking, we have introduced the `@sanity_function` and `@performance_function` decorators, which execute from the test's stage directory and they are called during the respective phases, so you avoid also any deferred expression evaluation headaches. If you write your test as follows it will be more robust:

```python
import reframe as rfm
import reframe.utility.sanity as sn
from reframe.core.backends import getlauncher


@rfm.simple_test
class HelloTest(rfm.RegressionTest):
    valid_systems = ['*']
    valid_prog_environs = ['*']
    sourcepath = 'hello.c'
    reference = {
        '*': {'Total Time': (0, None, None, 's')}
    }

    @run_before('run')
    def replace_launcher(self):
        self.job.launcher = getlauncher('local')()

    @sanity_function
    def validate_output(self):
        return sn.assert_found(r'Hello', self.stdout)

    @performance_function('s', perf_key='Total Time')
    def total_time(self):
        pref_regex = r'Hello\s(\S+)'
        times = sn.extractall(pref_regex, self.stdout, 1, int).evaluate()
        return sum(times) / len(times)
```

Regarding the error, I don't know if much can be done, because it is accurate; this is what happened, but it would help I guess if we reported the absolute path in the error message.

**Comment by OliverPerks** [comment]
Hey Vasileios, thanks for this - I really like this new format.
However, I have my suspicions that this is not quite the issue, or rather my understanding isn't quite right.
This function works fine if using `sn.extractsingle` type calls, but not the ones above, suggesting it is a referral problem.
Out of interest what are your thoughts on using absolute paths for these file using `self.stagedir` as the root?

**Comment by vkarak** [comment]
Hi Olly,

> This function works fine if using sn.extractsingle type calls, but not the ones above, suggesting it is a referral problem.

I guess you mean `sn.avg()` or any sanity function. If yes, then your example is expected to work! The reason lies in the implicit evaluation rules of the deferred expression. Generally, if a deferred expression is passed as an operand to any other expression, it will generate a new deferred expression, but there are three cases that its evaluation is triggered automatically:

1. When you try to iterate over it (i.e., include it in any sort of `for` loop)
2. When you try to take its truthy value, i.e., call explicitly or implicitly `bool()` upon it (i.e., include it as a condition in an `if` statement)
3. Whey you try to invoke `str()` on it (i.e., try to print it).

Your example falls into category (1): the builtin `sum()` will try to iterate over your deferrable, thus triggering its immediate evaluation. However, as I stated before, in your case this is isn't a big problem, because the `stdout` *is* produced since you're running in a pre-sanity hook. The problem is that the evaluation is triggered while you are not in the stage directory of your test. As a matter of fact, try to write your original example as follows:

```python
import reframe.utility.osext as osext


with osext.change_dir(self.stagedir):
    self.perf_patterns = {...}
```

It will work! (actually you need to replace `len()` with `sn.len()`, but that's not the point as you also said).

Now, why does this work if you use `sn.avg()` or even `sn.sum(times)/sn.len(times)`. The `sn.` variants of the builtins return a deferred expression instead of trying to iterate over their value and thus trigger an evaluation. So `sn.sum(times)` would be a deferred expression and so it will be the division. Now `perf_patterns`'s rhs value will be a deferred expression. The framework will evaluate this during the performance stage, but it will do so from within the stage directory of the test. That's why this works and not your original example. The mechanism of deferrable expressions and the rules of the implicit evaluation are detailed [here](http://localhost:8000/deferrables.html).

> Out of interest what are your thoughts on using absolute paths for these file using self.stagedir as the root?

I am now working on a PR that reports the absolute paths in such errors and I'm introducing some improvements in the documentation. And, in fact, I tried it with your example, and it shows clearly that it tries to find the `stdout` in the current working directory.
